api/src/main/java/org/apache/iceberg/SortOrder.java

```diff
@@ -0,0 +1,212 @@
+    if (anotherSortOrder.fields.length > fields.length) {
+      return false;
+    }
+
+    // this ordering has either more or the same number of sort fields
+    return IntStream.range(0, anotherSortOrder.fields.length)
+        .allMatch(index -> fields[index].equals(anotherSortOrder.fields[index]));
+  }
+
+  public boolean sameOrder(SortOrder anotherSortOrder) {
+    return Arrays.equals(fields, anotherSortOrder.fields);
+  }
+
+  private List<SortField> lazyFieldList() {
+    if (fieldList == null) {
+      synchronized (this) {
+        if (fieldList == null) {
+          this.fieldList = ImmutableList.copyOf(fields);
+        }
+      }
+    }
+    return fieldList;
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append("[");
+    for (SortField field : fields) {
+      sb.append("\n");
+      sb.append("  ").append(field);
+    }
+    if (fields.length > 0) {
+      sb.append("\n");
+    }
+    sb.append("]");
+    return sb.toString();
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    } else if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+
+    SortOrder that = (SortOrder) other;
+    return orderId == that.orderId && Arrays.equals(fields, that.fields);
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * Integer.hashCode(orderId) + Arrays.hashCode(fields);
+  }
+
+  public static SortOrder unsorted() {
+    return UNSORTED_ORDER;
+  }
+
+  public static Builder builderFor(Schema schema) {
+    return new Builder(schema);
+  }
+
+  public static class Builder {
+    private final Schema schema;
+    private final List<SortField> fields = Lists.newArrayList();
+    private int orderId = 0;
+    private boolean caseSensitive = true;
+
+    private Builder(Schema schema) {
+      this.schema = schema;
+    }
+
+    public Builder asc(String name, NullOrder nullOrder) {
+      return addSortField(Expressions.ref(name), SortDirection.ASC, nullOrder);
+    }
+
+    public Builder asc(Term term, NullOrder nullOrder) {
+      return addSortField(term, SortDirection.ASC, nullOrder);
+    }
+
+    public Builder desc(String name, NullOrder nullOrder) {
+      return addSortField(Expressions.ref(name), SortDirection.DESC, nullOrder);
+    }
+
+    public Builder desc(Term term, NullOrder nullOrder) {
+      return addSortField(term, SortDirection.DESC, nullOrder);
+    }
+
+    public Builder withOrderId(int newOrderId) {
+      this.orderId = newOrderId;
+      return this;
+    }
+
+    public Builder caseSensitive(boolean sortCaseSensitive) {
+      this.caseSensitive = sortCaseSensitive;
+      return this;
+    }
+
+    Builder addSortField(Term term, SortDirection direction, NullOrder nullOrder) {
+      Preconditions.checkArgument(term instanceof UnboundTerm, "Term must be unbound");
+      BoundTerm<?> boundTerm = ((UnboundTerm<?>) term).bind(schema.asStruct(), caseSensitive);
```

Comment:
Probably worth a comment that states `ValidationException` is thrown by `bind` if binding fails so we can assume that `boundTerm` is defined correctly.

Reply:
Will add.

Reply:
Done.